packages/hardhat-core/src/internal/vendor/await-semaphore/index.ts

```diff
@@ -64,3 +67,154 @@ export class Mutex extends Semaphore {
+      `Starting mutex process withy mutex file '${this._mutexFilePath}'`
+    );
+
+    while (true) {
+      if (await this._tryToAcquireMutex()) {
+        // Mutex has been acquired
+        return this._executeFunctionAndReleaseMutex(f);
+      }
+
+      // Mutex not acquired
+      if (this._isMutexFileTooOld()) {
+        // If the mutex file is too old, it likely indicates a deadlock, so the file should be removed
+        this._log(
+          `Current mutex file is too old, removing it at path '${this._mutexFilePath}'`
+        );
+        this._deleteMutexFile();
+      } else {
+        // wait
+        await this._waitMs();
+      }
+    }
+  }
+
+  private async _tryToAcquireMutex() {
+    try {
+      // Create a file only if it does not exist
+      fs.writeFileSync(this._mutexFilePath, "", { flag: "wx+" });
+      return true;
+    } catch (error: any) {
+      if (error.code === "EEXIST") {
+        // File already exists, so the mutex is already acquired
+        return false;
+      }
+
+      throw error;
+    }
+  }
+
+  private async _executeFunctionAndReleaseMutex<T>(
+    f: () => Promise<T>
+  ): Promise<T> {
+    this._log(`Mutex acquired at path '${this._mutexFilePath}'`);
+
+    // Set a timeout to ensure that the function does not take longer than the allowed time
+    let timeoutHandle: NodeJS.Timeout;
+    const timeoutChecker = new Promise<never>((_, reject) => {
+      timeoutHandle = setTimeout(() => {
+        reject(
+          new Error(
+            "The function's execution timed out while it had ownership of the mutex"
+          )
+        );
+      }, this._defaultTimeoutInMs);
+    });
+
+    try {
+      // Execute the function passed to the mutex, no other function will be able to run at the same time.
+      // If the function will take more than the timeout, it will be terminated with an error and the mutex will be released.
+      // Promise.race is used to race the timeoutChecker against the execution of fn().
+      // Whichever completes first will be the outcome of Promise.race.
+      return await Promise.race([f(), timeoutChecker]).then((result) => {
+        clearTimeout(timeoutHandle);
+
+        // Release the mutex
+        this._deleteMutexFile();
+
+        this._log(`Mutex released at path '${this._mutexFilePath}'`);
+
+        return result;
+      });
+    } catch (error: any) {
+      // Catch any error to avoid dead locks.
+      // Remove the mutex file and re-throw the error
+      this._deleteMutexFile();
+      throw error;
+    }
+  }
+
+  private _isMutexFileTooOld(): boolean {
+    const now = new Date();
+
+    let fileStat;
+    try {
+      fileStat = fs.statSync(this._mutexFilePath);
+    } catch (error: any) {
+      if (error.code === "ENOENT") {
+        // The file might have been deleted by another process while this function was trying to access it.
+        return false;
+      }
+
+      throw error;
+    }
+
+    const fileDate = new Date(fileStat.mtime);
```

Comment:
Although for this particular case is not super relevant, in general you want to use `ctime` instead of `mtime`, because `mtime` is only about the file contents being changed, while `ctime` also includes metadata changes.

Reply:
Done